assets/migrations/postgres/004_new_user_parts_columns.sql

```diff
@@ -0,0 +1,114 @@
+-- +goose NO TRANSACTION
+-- This migration adds 3 more columns to the tuple and changelog tables, and fills past and future data with a trigger.
+-- This migration also adds more indexes to the tuple table. It does so concurrently so that the main table isn't locked
+
+-- +goose Up
+
+ALTER TABLE tuple
+    ADD COLUMN IF NOT EXISTS user_object_type TEXT,
+    ADD COLUMN IF NOT EXISTS user_object_id TEXT,
+    ADD COLUMN IF NOT EXISTS user_relation TEXT;
+
+-- jon becomes (..., jon, ...)
+-- user:jon becomes (user, jon, ...)
+-- team:* becomes (team, *, ...)
+-- group:eng#member becomes (group, eng, member)
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION set_new_columns() RETURNS TRIGGER AS $$
+BEGIN
+        NEW.user_object_type = (CASE
+                                    WHEN position(':' in NEW._user) > 0 THEN split_part(NEW._user, ':', 1)
+                                    ELSE ''
+            END),
+
+        NEW.user_object_id = (CASE
+                                  WHEN position('#' in NEW._user) > 0 THEN split_part(split_part(NEW._user, ':', -1), '#', 1)
+                                  ELSE split_part(NEW._user, ':', -1)
+            END),
+
+        NEW.user_relation = (CASE
+                                 WHEN position('#' in NEW._user) > 0 THEN split_part(NEW._user, '#', -1)
+                                 ELSE ''
+            END);
+END;
+$$ language 'plpgsql';
+-- +goose StatementEnd
+
+UPDATE tuple SET
+     user_object_type = (CASE
+                             WHEN position(':' in _user) > 0 THEN split_part(_user, ':', 1)
+                             ELSE ''
+         END),
+     user_object_id = (CASE
+                           WHEN position('#' in _user) > 0 THEN split_part(split_part(_user, ':', -1), '#', 1)
+                           ELSE split_part(_user, ':', -1)
+         END),
+     user_relation = (CASE
+                          WHEN position('#' in _user) > 0 THEN split_part(_user, '#', -1)
+                          ELSE ''
+         END);
+
+CREATE TRIGGER migrate_user_column
+    BEFORE UPDATE ON tuple
```

Comment:
Should be `BEFORE INSERT ON tuple`

If we want to be extra sure we should

Reply:
See https://github.com/openfga/openfga/pull/702/files#r1203022455 - same issue with the 'changelog' table.

Reply:
the `OR UPDATE` is unnecessary, no? 🤔

Reply:
It is, but it covers our bases in case of unforseen migrations in the future. Doesn't hurt to add, because it if the database never sees an UPDATE it has no effect.